File: macos/Sources/ElevateCore/Support/PanelStatus.swift
import Foundation

/// What the menu bar label needs to know, derived from the assignments alone.
public struct PanelStatus: Equatable, Sendable {
    public let activeCount: Int
    public let expiringSoon: Bool
    public let pendingApproval: Bool

    public init(activeCount: Int, expiringSoon: Bool, pendingApproval: Bool) {
        self.activeCount = activeCount
        self.expiringSoon = expiringSoon
        self.pendingApproval = pendingApproval
    }

    /// `expiringSoon` is true when an active assignment ends within `soonWithin` seconds of `now` (and has not ended yet).
    public static func compute(_ assignments: [ActiveAssignment], now: Date, soonWithin: TimeInterval = 300) -> PanelStatus {
        var active = 0, soon = false, pending = false
        for a in assignments {
            switch a.status {
            case .active:
                active += 1
                if let end = a.endDateTime {
                    let left = end.timeIntervalSince(now)
                    if left > 0 && left <= soonWithin { soon = true }
                }
            case .pendingApproval: pending = true
            case .pendingProvisioning, .failed: break
            }
        }
        return PanelStatus(activeCount: active, expiringSoon: soon, pendingApproval: pending)
    }
}

File: windows/src/Elevate.Core/Support/PanelStatus.cs
using Elevate.Core.Models;

namespace Elevate.Core.Support;

/// <summary>What the tray icon needs to know, derived from the assignments alone.</summary>
public readonly record struct PanelStatus(int ActiveCount, bool ExpiringSoon, bool PendingApproval)
{
    public static readonly TimeSpan DefaultSoonWithin = TimeSpan.FromMinutes(5);

    /// <summary>
    /// <see cref="ExpiringSoon"/> is true when an active assignment ends within <paramref name="soonWithin"/>
    /// of <paramref name="now"/> (and has not ended yet).
    /// </summary>
    public static PanelStatus Compute(IEnumerable<ActiveAssignment> assignments, DateTimeOffset now, TimeSpan? soonWithin = null)
    {
        ArgumentNullException.ThrowIfNull(assignments);
        var within = soonWithin ?? DefaultSoonWithin;
        var active = 0;
        var soon = false;
        var pending = false;
        foreach (var a in assignments)
        {
            switch (a.Status.Kind)
            {
                case AssignmentStatusKind.Active:
                    active += 1;
                    if (a.EndDateTime is { } end)
                    {
                        var left = end - now;
                        if (left > TimeSpan.Zero && left <= within)
                        {
                            soon = true;
                        }
                    }

                    break;
                case AssignmentStatusKind.PendingApproval:
                case AssignmentStatusKind.Scheduled:
                    pending = true;
                    break;
                default:
                    break;
            }
        }

        return new PanelStatus(active, soon, pending);
    }
}
